<!doctype html>
<html class="no-js" lang="en">

<head>
  <!-- Global site tag (gtag.js) - Google Analytics -->
  <script async src="https://www.googletagmanager.com/gtag/js?id=UA-134931760-1"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());

    gtag('config', 'UA-134931760-1');
  </script>

  <meta charset="utf-8">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <title>MS SQL 2011 - Error Handling</title>
  <link rel="icon" type="image/x-icon" href="/favicon.ico" />
  <meta name="description" content="A new useful addition for SQL Server 2011 (Denali) is Throw expression. The developers on .Net probably already guessed where and how it will be used....">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <link rel="stylesheet" href="../../css/main.css">

  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  <script>window.jQuery || document.write('<script src="../../js/vendors/jquery-3.3.1.min.js"><\/script>')</script>

  <script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
  <script>
       (adsbygoogle = window.adsbygoogle || []).push({
            google_ad_client: "ca-pub-6974184241884155",
            enable_page_level_ads: true
       });
  </script>
</head>

<body>
  <!--[if lte IE 9]>
    <p class="browserupgrade">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience and security.</p>
  <![endif]-->
  <header class="page-header js-page-header">
    <a class="page-header-logo-container" href="https://weekly-geekly.github.io/index.html"></a>
    <div class="page-header-text">Geekly Articles each Day</div>
  </header>
  <nav class="page-headings-container js-page-headings-container"></nav>
  <div class="tools-bar js-tools-bar">
    <!-- <a href="../../search.html" title="Search">üîé</a> -->
    <a class="js-list-of-headings-button" data-state="closed" href="#" title="Headings">üìú</a>
    <a class="js-go-to-top-button" href="#" title="Go to Top">‚¨ÜÔ∏è</a>
    <a class="js-go-to-bottom-button" href="#" title="Go to Bottom">‚¨áÔ∏è</a>
  </div>
  <a href="http://bit.ly/donateToWeeklyGeekly" class="donate-btn">DONATE</a>
  <section class="page js-page"><h1>MS SQL 2011 - Error Handling</h1><div class="post__text post__text-html js-mediator-article">  A new useful addition for SQL Server 2011 (Denali) is <strong>Throw</strong> expression.  The developers on .Net probably already guessed where and how it will be used. <br><br>  This word can be used in conjunction with the Try ... Catch control structure and allows you to send a notification about the occurrence of a runtime error.  When an exception occurs, the program looks for the nearest Catch block that can handle the exception.  Using this expression inside the Catch block, you can change the error output.  Moreover, now it is possible to raise an exception arbitrarily anywhere in the script. <br><br>  Next, consider the various ways of catching the exceptions that SQL Server provides from version 2000 to version 2011, with an indication of the pros and cons. 
      <br>
        <script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
        <ins class="adsbygoogle"
          style="display:block; text-align:center;"
          data-ad-layout="in-article"
          data-ad-format="fluid"
          data-ad-client="ca-pub-6974184241884155"
          data-ad-slot="8945601208"></ins>
        <script>
          (adsbygoogle = window.adsbygoogle || []).push({});
        </script>
      <br>
    <a name="habracut"></a><br><br>  For all cases considered, the <em>tbl_ExceptionTest</em> table will be used. <br><br> <a href=""><img title="01" src="https://habrastorage.org/getpro/habr/post_images/559/f66/ed5/559f66ed5deb82355466c7e9149ce00c.jpg" width="388" height="94"></a> <br><br>  In order not to pierce the designer with the mouse, you can run the following script to create the desired table (generated automatically). <br><pre><code class="sql hljs">IF EXISTS (<span class="hljs-keyword"><span class="hljs-keyword">SELECT</span></span> * <span class="hljs-keyword"><span class="hljs-keyword">FROM</span></span> sys.objects <span class="hljs-keyword"><span class="hljs-keyword">WHERE</span></span> <span class="hljs-keyword"><span class="hljs-keyword">name</span></span> = <span class="hljs-string"><span class="hljs-string">'tbl_ExceptionTest'</span></span> <span class="hljs-keyword"><span class="hljs-keyword">AND</span></span> <span class="hljs-keyword"><span class="hljs-keyword">type</span></span> = <span class="hljs-string"><span class="hljs-string">'U'</span></span>) <span class="hljs-keyword"><span class="hljs-keyword">DROP</span></span> <span class="hljs-keyword"><span class="hljs-keyword">TABLE</span></span> tbl_ExceptionTest <span class="hljs-keyword"><span class="hljs-keyword">GO</span></span> <span class="hljs-keyword"><span class="hljs-keyword">SET</span></span> ANSI_NULLS <span class="hljs-keyword"><span class="hljs-keyword">ON</span></span> <span class="hljs-keyword"><span class="hljs-keyword">GO</span></span> <span class="hljs-keyword"><span class="hljs-keyword">SET</span></span> QUOTED_IDENTIFIER <span class="hljs-keyword"><span class="hljs-keyword">ON</span></span> <span class="hljs-keyword"><span class="hljs-keyword">GO</span></span> <span class="hljs-keyword"><span class="hljs-keyword">CREATE</span></span> <span class="hljs-keyword"><span class="hljs-keyword">TABLE</span></span> [dbo].[tbl_ExceptionTest]( [<span class="hljs-keyword"><span class="hljs-keyword">Id</span></span>] [<span class="hljs-built_in"><span class="hljs-built_in">int</span></span>] <span class="hljs-keyword"><span class="hljs-keyword">IDENTITY</span></span>(<span class="hljs-number"><span class="hljs-number">1</span></span>,<span class="hljs-number"><span class="hljs-number">1</span></span>) <span class="hljs-keyword"><span class="hljs-keyword">NOT</span></span> <span class="hljs-literal"><span class="hljs-literal">NULL</span></span>, [Phone <span class="hljs-built_in"><span class="hljs-built_in">Number</span></span>] [<span class="hljs-built_in"><span class="hljs-built_in">int</span></span>] <span class="hljs-keyword"><span class="hljs-keyword">NOT</span></span> <span class="hljs-literal"><span class="hljs-literal">NULL</span></span>, <span class="hljs-keyword"><span class="hljs-keyword">CONSTRAINT</span></span> [PK_tbl_ExceptionTest] PRIMARY <span class="hljs-keyword"><span class="hljs-keyword">KEY</span></span> CLUSTERED )</code> </pre> <br>  Next, we will try to add several entries to the table and generate exceptions when entering inappropriate data in the Phone Number column. <br><br><h2>  Error Handling in SQL Server 2000 (Sphinx) </h2><br><h3>  Using the global variable @@ ERROR </h3><br>  Returning at the time of using SQL Server 2000, we recall that using the @@ Error variable was at that time the most progressive and efficient way to handle errors.  This variable was responsible for returning the integer value of the error that occurred in the last expression executed.  The error value could be both positive and negative, only 0 indicated the success of the operation.  The value of the variable changed after each expression executed. <br><br>  Let's look at using @@ Error in action. <br><pre> <code class="sql hljs"><span class="hljs-comment"><span class="hljs-comment">--   #tblExceptionTest  ,   . If OBJECT_ID('tempdb..#tblExceptionTest') Is not null Begin Drop Table #tblExceptionTest End --    #tblExceptionTest Create Table #tblExceptionTest (Id int identity, [Phone Number] varchar(10) not null) --  Begin Transaction TranExcp__2000_@@Error --   --       @@ERROR Declare @ErrorNum int --      Declare @i int --   Set @i =1 --   While(@i &lt;= 4) Begin --       null   Phone Number If(@i = 4) Begin Insert into #tblExceptionTest([Phone Number]) Values(null) Set @ErrorNum = @@ERROR End Else --      Begin Insert into #tblExceptionTest([Phone Number]) Values(cast(@i as varchar(2)) + '12345678') End Set @i = @i +1 End --  while --   ,      If @ErrorNum &lt;&gt; 0 Begin Rollback Transaction TranExcp__2000_@@Error --      RAISERROR ('Attempt to insert null value in [Phone Number] is not allowed',16,1) End --   Else If @ErrorNum = 0 Begin Commit Transaction TranExcp__2000_@@Error End --   Select * from #tblExceptionTest</span></span></code> </pre> <br>  The general meaning of the script is to ensure that in the last entry we intentionally cause an error and read its value from a local variable.  If the error value is not zero, then we show a sensible warning to the user.  If there are no errors, then save the results. <br><br>  Running this script will result in an error, as shown below. <br><br>  <strong><em>Msg 515, level 16, state 2, line 26</em></strong>  <strong><em>column does not allow nulls.</em></strong>  <strong><em>INSERT fails.</em></strong>  <strong><em>The statement has been terminated.</em></strong>  <strong><em>Msg 50000, Level 16, State 1, Line 43 Attempt to insert null value in [Phone Number] is not allowed</em></strong> <br><br>  Naturally, the entire transaction will be rolled back and nothing will be entered into the table. <br><br><h3>  Disadvantages of the @@ Error approach </h3><br><ul><li>  The value of the @@ Error variable should be checked <strong>immediately</strong> after the execution of the request / command. </li><li>  Since @@ Error is constantly changing, we are forced to set up a separate variable to save and display the error code. </li><li>  Together with a special error message pointing to the logical meaning of the error, technical information is displayed that is not interesting to users. </li></ul><br>  If you want to know more details and nuances of using @@ Error, then I advise you to refer to the article about <a href="http://msdn.microsoft.com/en-us/library/aa933181(v%3Dsql.80).aspx">@@ Error</a> . <br><br><h3>  Using the global @@ TRANCOUNT variable </h3><br>  This variable returns the number of transactions executed at the time the variable was accessed.  From the description it is already clear that it is approximately the same as @@ ERROR, i.e.  constantly changing during the execution of transactions.  This again leads us to use local variables to store values ‚Äã‚Äãat the point in time. <br><br>  Each BEGIN TRANSACTION call increases the @@ TRANCOUNT value by 1 and each COMMIT TRANSACTION call decreases its value by 1. ROLLBACK TRANSACTION does not change the @@ TRANCOUNT value.  Entries are considered to be made only when the @@ TRANCOUNT value reaches 0. <br><br>  Consider using @@ TRANCOUNT in the following example. <br><pre> <code class="sql hljs"><span class="hljs-comment"><span class="hljs-comment">--   #tblExceptionTest ,    If OBJECT_ID('tempdb..#tblExceptionTest') Is not null Begin Drop Table #tblExceptionTest End --    Create Table #tblExceptionTest (Id int identity, [Phone Number] varchar(10) not null) --   Begin Transaction TranExcp__2000_@@TRANCOUNT --  --     @@TRANCOUNT Declare @TransactionCount int --  Declare @i int --   Set @i =1 --   While(@i &lt;= 4) Begin --       null   Phone Number If(@i = 4) Begin Insert into #tblExceptionTest([Phone Number]) Values(null) Set @TransactionCount = @@TRANCOUNT End Else --      Begin Insert into #tblExceptionTest([Phone Number]) Values(cast(@i as varchar(2)) + '12345678') End Set @i = @i +1 End --  while --   ,        If @TransactionCount &lt;&gt; 0 Begin Rollback Transaction TranExcp__2000_@@TRANCOUNT --    RAISERROR ('Attempt to insert null value in [Phone Number] is not allowed',16,1) End --   Else If @TransactionCount = 0 Begin Commit Transaction TranExcp__2000_@@TRANCOUNT End --   Select * from #tblExceptionTest</span></span></code> </pre> <br>  In this script, we rely on the number of closed transactions.  Transactions can be nested, so this method has the right to exist. <br><br>  For more information on <a href="http://msdn.microsoft.com/en-us/library/ms187967.aspx">@@ TRANCOUNT,</a> refer to MSDN. <br><br><h3>  Using the global variable @@ ROWCOUNT </h3><br>  This variable returns the number of rows changed as a result of the query / command. <br><br>  The behavior is the same as the previous two, so we save the intermediate results in a local variable for later analysis. <br><br>  Example: <br><pre> <code class="sql hljs">If OBJECT_ID('tempdb..<span class="hljs-comment"><span class="hljs-comment">#tblExceptionTest') Is not null Begin Drop Table #tblExceptionTest End Create Table #tblExceptionTest (Id int identity, [Phone Number] varchar(10) not null) Begin Transaction TranExcp__2000_@@ROWCOUNT Save Transaction TranExcp__SavePoint Declare @RowCount int Declare @i int Set @i =1 While(@i &lt;= 4) Begin If(@i = 4) Begin Insert into #tblExceptionTest([Phone Number]) Values(null) Set @RowCount = @@ROWCOUNT End Else Begin Insert into #tblExceptionTest([Phone Number]) Values(cast(@i as varchar(2)) + '12345678') End Set @i = @i +1 End If @RowCount = 0 Begin Rollback Transaction TranExcp__SavePoint RAISERROR ('Attempt to insert null value in [Phone Number] is not allowed',16,1) End Else If @RowCount &lt;&gt; 0 Begin Commit Transaction TranExcp__2000_@@ROWCOUNT End Select * from #tblExceptionTest</span></span></code> </pre> <br>  In this case, we expect one record to be inserted into the table, but if the number of records inserted is zero, then obviously something is wrong. <br>  For more details on using <a href="http://technet.microsoft.com/en-us/library/ms187316.aspx">@@ ROWCOUNT</a> read MSDN. <br><br><h2>  Error Handling in SQL Server 2005/2008 (Yukon / Katmai) </h2><br>  After the introduction of SQL Server 2005 to the market and the development of its ideas in SQL Server 2008, TSql developers got a new Try ... Catch block.  Now it is possible to catch exceptions without losing the transactional context. <br><br>  An example of using the Try ... Catch block. <br><pre> <code class="sql hljs">If OBJECT_ID('tempdb..<span class="hljs-comment"><span class="hljs-comment">#tblExceptionTest') Is not null Begin Drop Table #tblExceptionTest End Begin TRY Create Table #tblExceptionTest (Id int identity, [Phone Number] varchar(10) not null) Begin Transaction TranExcpHandlingTest_2005_2008 Declare @i int Set @i =1 While(@i &lt;= 4) Begin If(@i = 4) Begin Insert into #tblExceptionTest([Phone Number]) Values(null) End Else Begin Insert into #tblExceptionTest([Phone Number]) Values(cast(@i as varchar(2)) + '12345678') End Set @i = @i +1 End Commit Transaction TranExcpHandlingTest_2005_2008 End Try Begin Catch Begin Rollback Transaction TranExcpHandlingTest_2005_2008 RAISERROR ('Attempt to insert null value in [Phone Number] is not allowed',16,1) End End Catch Select * From #tblExceptionTest</span></span></code> </pre> <br>  The example no longer uses auxiliary variables to determine the script execution error based on indirect signs. <br><br>  After starting the script, we will receive a message of the following form: <br><br>  <strong><em>Msg 50000, Level 16, State 1, Line 45 Attempt to insert null value in [Phone Number] is not allowed</em></strong> <br><br>  As you may have noticed, this time only what was specified in the error message was displayed.  No additional, embarrassing user messages, SQL Server showed.  Executable code is framed in a try block and error handling in a catch block.  The code turns out pure and clear for understanding.  If all the desired code has passed without errors, the code from the Catch block will not be called. <br><br>  The most important thing is that the Catch block represents a set of functions for a detailed analysis of the causes of the error and the ability to inform the user at the proper level.  Functions to parse an exception: <br><ul><li>  ERROR_NUMBER </li><li>  ERROR_SEVERITY </li><li>  ERROR_STATE </li><li>  ERROR_LINE </li><li>  ERROR_PROCEDURE </li><li>  ERROR_MESSAGE </li></ul><br>  Using these functions, we will try to rewrite the Catch script block, which would be presented before. <br><pre> <code class="sql hljs"><span class="hljs-keyword"><span class="hljs-keyword">Begin</span></span> Catch <span class="hljs-comment"><span class="hljs-comment">--   Begin --   Rollback Transaction TranExcpHandlingTest_2005_2008 SELECT ERROR_NUMBER() AS ErrorNumber, ERROR_SEVERITY() AS ErrorSeverity, ERROR_STATE() AS ErrorState, ERROR_PROCEDURE() AS ErrorProcedure, ERROR_LINE() AS ErrorLine, ERROR_MESSAGE() AS ErrorMessage; End End Catch</span></span></code> </pre> <br>  Now we get the following response from the server: <br><br> <a href=""><img title="02" src="https://habrastorage.org/getpro/habr/post_images/e79/d21/9e8/e79d219e848e3929045c857fed7d2102.jpg" width="572" height="44"></a> <br><br><h3>  Disadvantages of using the RaiseError function </h3><br>  1 If we recall what this function caused in the Catch block showed, then we note that it referred to line number 45 as the source of the problems. <br><br> <a href=""><img title="03" src="https://habrastorage.org/getpro/habr/post_images/c69/8f3/a9f/c698f3a9f0f771923a971f2bb0c896d4.jpg" width="449" height="55"></a> <br><br>  However, in reality, the error occurred in line number 24, where it was written <br><br>  <strong>Insert into #tblExceptionTest ([Phone Number]) Values ‚Äã‚Äã(null)</strong> <br><br>  While the ERROR_LINE () function always returns the actual location of the error.  Another way to show how new features work is: <br><pre> <code class="sql hljs"><span class="hljs-keyword"><span class="hljs-keyword">Begin</span></span> Catch <span class="hljs-keyword"><span class="hljs-keyword">Begin</span></span> <span class="hljs-keyword"><span class="hljs-keyword">Rollback</span></span> <span class="hljs-keyword"><span class="hljs-keyword">Transaction</span></span> TranExcpHandlingTest_2005_2008 <span class="hljs-keyword"><span class="hljs-keyword">DECLARE</span></span> @errNumber <span class="hljs-built_in"><span class="hljs-built_in">INT</span></span> = ERROR_NUMBER() <span class="hljs-keyword"><span class="hljs-keyword">DECLARE</span></span> @errMessage <span class="hljs-built_in"><span class="hljs-built_in">VARCHAR</span></span>(<span class="hljs-number"><span class="hljs-number">500</span></span>) = <span class="hljs-string"><span class="hljs-string">'Attempt to insert null value in [Phone Number] is not allowed'</span></span> RAISERROR(<span class="hljs-string"><span class="hljs-string">'Error Number: %d, Message: %s'</span></span>, <span class="hljs-number"><span class="hljs-number">16</span></span>, <span class="hljs-number"><span class="hljs-number">1</span></span>, @errNumber, @errMessage) <span class="hljs-keyword"><span class="hljs-keyword">End</span></span> <span class="hljs-keyword"><span class="hljs-keyword">End</span></span> Catch</code> </pre> <br><br><br>  In this case, the SQL Server engine will display the following message: <br><br> <a href=""><img title="04" src="https://habrastorage.org/getpro/habr/post_images/0f8/647/21f/0f864721f8bbb8b5b5c3c323573a420e.jpg" width="605" height="40"></a> <br><br>  From which we can conclude that the use of RaiseError makes it impossible to point to the real place in the script where the exception occurred. <br><br>  2 The next drawback of the RaiseError function is that it is not possible to re-trigger the same exception itself, to pass up the call hierarchy.  So, if you rewrite the Catch block as shown below <br><pre> <code class="sql hljs"><span class="hljs-keyword"><span class="hljs-keyword">Begin</span></span> Catch <span class="hljs-keyword"><span class="hljs-keyword">Begin</span></span> <span class="hljs-keyword"><span class="hljs-keyword">Rollback</span></span> <span class="hljs-keyword"><span class="hljs-keyword">Transaction</span></span> TranExcpHandlingTest_2005_2008 RAISERROR(<span class="hljs-number"><span class="hljs-number">515</span></span>, <span class="hljs-number"><span class="hljs-number">16</span></span>, <span class="hljs-number"><span class="hljs-number">1</span></span>) <span class="hljs-keyword"><span class="hljs-keyword">End</span></span> <span class="hljs-keyword"><span class="hljs-keyword">End</span></span> Catch</code> </pre> <br>  That error message received will be: <br><br>  <strong><em>Msg 2732, Level 16, State 1, Line 46 Error number 515 is invalid.</em></strong>  <strong><em>The number must be from 13000 through 2147483647 and it cannot be 50000</em></strong> <br><br>  The reason for this is that in order to initiate a new error message, the error number must be contained in the <strong>sys</strong> table <strong>.</strong>  <strong>messages</strong> . <br><br>  For a more detailed study of the function RaiseError, it is recommended to read: <br><ul><li>  <a href="http://www.codeproject.com/KB/database/Raiserror.aspx">Detailed look at RaiseError in SQL Server 2005</a> </li><li>  <a href="http://msdn.microsoft.com/en-us/library/ms177497.aspx">Use RaiseError</a> </li></ul><br><br><h2>  Error Handling in SQL Server 2011 (Denali) </h2><br>  The above disadvantages of the RaiseError function can be successfully overcome with the <strong>new</strong> <strong>Throw</strong> <strong>command</strong> . <br><br>  The first drawback of the RaiseError function, which we pointed out earlier, is the inability to refer to the exact error string.  Consider how far from the location of the error we find ourselves when using the Throw command. <br><br>  Rewrite the Catch block using the Throw command. <br><pre> <code class="sql hljs"><span class="hljs-keyword"><span class="hljs-keyword">Begin</span></span> Catch <span class="hljs-keyword"><span class="hljs-keyword">Begin</span></span> <span class="hljs-keyword"><span class="hljs-keyword">Rollback</span></span> <span class="hljs-keyword"><span class="hljs-keyword">Transaction</span></span> TranExcpHandlingTest_2011; THROW <span class="hljs-keyword"><span class="hljs-keyword">End</span></span> <span class="hljs-keyword"><span class="hljs-keyword">End</span></span> Catch</code> </pre> <br>  The output will be: <br><br> <a href=""><img title="05" src="https://habrastorage.org/getpro/habr/post_images/201/b88/ac6/201b88ac64b487b04f0f7d95d86fdca1.jpg" width="605" height="57"></a> <br><br>  This is exactly where the error occurred.  Well, it works perfectly well for now. <br><br>  The second disadvantage was that the RaiseError function could not re-trigger the exception because RAISE ERROR expects an error number, which is stored in the sys.messages table.  The Throw command does not expect that the error number should be from the sys.messages system table range, but the number can be set from the range from 50,000 to 2,147,483,647 inclusive. <br><br>  Again we will change the Catch block according to new knowledge. <br><pre> <code class="sql hljs"><span class="hljs-keyword"><span class="hljs-keyword">Begin</span></span> Catch <span class="hljs-keyword"><span class="hljs-keyword">Begin</span></span> <span class="hljs-keyword"><span class="hljs-keyword">Rollback</span></span> <span class="hljs-keyword"><span class="hljs-keyword">Transaction</span></span> TranExcpHandlingTest_2011; THROW 50001,'Attempt to <span class="hljs-keyword"><span class="hljs-keyword">insert</span></span> <span class="hljs-literal"><span class="hljs-literal">null</span></span> <span class="hljs-keyword"><span class="hljs-keyword">value</span></span> <span class="hljs-keyword"><span class="hljs-keyword">in</span></span> [Phone <span class="hljs-built_in"><span class="hljs-built_in">Number</span></span>] <span class="hljs-keyword"><span class="hljs-keyword">is</span></span> <span class="hljs-keyword"><span class="hljs-keyword">not</span></span> allowed<span class="hljs-string"><span class="hljs-string">',1 End End Catch</span></span></code> </pre> <br>  The result of the exception will be <br><br>  <strong><em>Msg 50001, Level 16, State 1, Line 45 Attempt to insert null value in [Phone Number] is not allowed</em></strong> <br><br>  Currently, SQL Server provides many ways to catch errors, but so far not all errors can be caught using the Try ... Catch block.  For example: <br><ul><li>  Syntax errors are captured by the query editor in SSMS </li><li>  Wrong object names </li></ul><br>  If you try to submit for execution the following script: <br><pre> <code class="sql hljs"><span class="hljs-keyword"><span class="hljs-keyword">Begin</span></span> Try <span class="hljs-comment"><span class="hljs-comment">--   tblInvalid Insert Into tblInvalid(Id,DOB) Values(1,DATEADD(year,1,'2011-02-26')) End Try Begin Catch --  THROW End Catch</span></span></code> </pre> <br>  We will receive the following error message: <br><br>  <strong><em>Msg 208, Level 16, State 0, Line 3 Invalid object name 'tblInvalid'.</em></strong> <br><br>  It turns out that it is almost impossible to intercept these types of errors. <br><br>  But.  As always, there is a little trick to do what you want.  The basic idea is to make two stored procedures and call one of the other in the Try ... Catch block and catch an exception.  To prove our assumption, we use the following script for experiments. <br><pre> <code class="sql hljs"><span class="hljs-comment"><span class="hljs-comment">--   ,  ,   If Exists (Select * from sys.objects where name = 'usp_InternalStoredProc' and type = 'P') Drop Procedure usp_InternalStoredProc Go --     Create Procedure usp_InternalStoredProc As Begin Begin Transaction TranExcpHandlingTest_2011 Begin Try --     Insert Into tblInvalid(Id,DOB) Values(1,DATEADD(year,1,'2011-02-26')) --   Commit Transaction TranExcpHandlingTest_2011 End Try Begin Catch If @@TRANCOUNT &gt; 0 Rollback Transaction TranExcpHandlingTest_2011 Print 'In catch block of internal stored procedure.... throwing the exception'; --   THROW End Catch End Go --       --   ,  ,   If Exists (Select * from sys.objects where name = 'usp_ExternalStoredProc' and type = 'P') Drop Procedure usp_ExternalStoredProc Go --     Create Procedure usp_ExternalStoredProc As Begin Begin Try --    Exec usp_InternalStoredProc End Try Begin Catch Print 'In catch block of external stored procedure.... throwing the exception'; SELECT ERROR_NUMBER() AS ErrorNumber ,ERROR_SEVERITY() AS ErrorSeverity ,ERROR_STATE() AS ErrorState ,ERROR_PROCEDURE() AS ErrorProcedure ,ERROR_LINE() AS ErrorLine ,ERROR_MESSAGE() AS ErrorMessage; THROW End Catch End Go --    Exec usp_ExternalStoredProc</span></span></code> </pre> <br>  When you run the <strong>ExternalStoredProc</strong> procedure, we get the message: <br><pre> <code class="sql hljs">In catch block of external stored procedure.... throwing the exception (1 row(s) affected) Msg 208, Level 16, State 1, Procedure usp_InternalStoredProc, Line 8 Invalid object name 'tblInvalid'.</code> </pre> <br>  And the Result panel will display the following data: <br><br> <a href=""><img title="06" src="https://habrastorage.org/getpro/habr/post_images/165/5fe/dab/1655fedabffd8140dab4b06f4eec10c9.jpg" width="606" height="67"></a> <br><br>  What we needed! <br><br>  Now a little explanation of how the code works.  We have 2 stored procedures: <em>usp_InternalStoredProc</em> and <em>usp_ExternalStoredProc</em> .  In <em>usp_InternalStoredProc,</em> we are trying to insert a record into the nonexistent table #tblInnerTempTable, as a result of which we get an exceptional situation, which in turn is caught by an external Catch block located in an external procedure. <br><br>  Moreover, the line and the text of the error fully meet our expectations and indicate the exact place. <br><br>  It is very important not to forget to close with a semicolon the expression before the THROW statement in the external procedure.  THROW must be a new command set.  Otherwise, get an error <br><br>  <strong><em>Incorrect syntax near 'THROW'.</em></strong> <br><br>  More detailed information about <a href="http://msdn.microsoft.com/en-us/library/ee677615(SQL.110).aspx">THROW</a> can be found in MSDN. <br><br>  <i>Transfers from the cycle:</i> <i><br></i>  <i>MS SQL Server 2011: <a href="http://habrahabr.ru/blogs/htranslations/123345/">Autonomous databases</a> , <a href="http://habrahabr.ru/blogs/sql/123446/">new Sequence object</a> , <a href="http://habrahabr.ru/blogs/sql/123491/">Offset operator</a> , <a href="http://habrahabr.ru/blogs/sql/123507/">error handling</a> , <a href="http://habrahabr.ru/blogs/sql/123573/">With Result Set construction</a> , <a href="http://habrahabr.ru/blogs/sql/123664/">new in SSMS</a> .</i> </div><p>Source: <a href="https://habr.com/ru/post/123507/">https://habr.com/ru/post/123507/</a></p>
<section class="more-articles-navigation-panel js-more-articles-navigation-panel">
<h4>More articles:</h4>
<nav class="list-of-articles-container js-list-of-articles-container"><ul class="list-of-pages js-list-of-pages">
<li><a href="../123491/index.html">MS SQL 2011 - New Offset Operator</a></li>
<li><a href="../123496/index.html">Mashable: Google abandons Blogger and Picasa brands in favor of Google+</a></li>
<li><a href="../123501/index.html">Pdf.js passed the first pixel test</a></li>
<li><a href="../123502/index.html">Using MongoDB in Java EE 6</a></li>
<li><a href="../123504/index.html">Fun with Google's Street View</a></li>
<li><a href="../123508/index.html">Generation of analytical surfaces on the example of maps</a></li>
<li><a href="../123509/index.html">Hot summer room UserAndLINUX ‚Ññ11</a></li>
<li><a href="../123510/index.html">Orbduino - mechanical arm controlled via the Internet</a></li>
<li><a href="../123513/index.html">Twitter is already valued at $ 7 billion.</a></li>
<li><a href="../123515/index.html">Useful Asterisk</a></li>
</ul></nav>
</section><br />
<a href="../../allArticles.html"><strong>All Articles</strong></a>
<script src="../../js/main.js"></script>

<!-- Yandex.Metrika counter -->
<script type="text/javascript" >
  (function (d, w, c) {
      (w[c] = w[c] || []).push(function() {
          try {
              w.yaCounter52496797 = new Ya.Metrika({
                  id:52496797,
                  clickmap:true,
                  trackLinks:true,
                  accurateTrackBounce:true,
                  webvisor:true
              });
          } catch(e) { }
      });

      var n = d.getElementsByTagName("script")[0],
          s = d.createElement("script"),
          f = function () { n.parentNode.insertBefore(s, n); };
      s.type = "text/javascript";
      s.async = true;
      s.src = "https://mc.yandex.ru/metrika/watch.js";

      if (w.opera == "[object Opera]") {
          d.addEventListener("DOMContentLoaded", f, false);
      } else { f(); }
  })(document, window, "yandex_metrika_callbacks");
</script>
<noscript><div><img src="https://mc.yandex.ru/watch/52496797" style="position:absolute; left:-9999px;" alt="" /></div></noscript>

<!-- Google Analytics -->
  <script>
    window.ga = function () { ga.q.push(arguments) }; ga.q = []; ga.l = +new Date;
    ga('create', 'UA-134931760-1', 'auto'); ga('send', 'pageview')
  </script>
  <script src="https://www.google-analytics.com/analytics.js" async defer></script>

</section>

  <footer class="page-footer">
    <div class="page-footer-legal-info-container page-footer-element">
      <p>
        Weekly-Geekly | <span class="page-footer-legal-info-year js-page-footer-legal-info-year">2019</span>
      </p>
    </div>
    <div class="page-footer-counters-container page-footer-element">
      <a class="page-footer-counter-clustrmap" href='#'  title='Visit tracker'><img src='https://clustrmaps.com/map_v2.png?cl=698e5a&w=271&t=t&d=6iCFw7uJz0zcOaoxz5k5PcLCJUzv2WG8G5V8M3U6Rc4&co=3a3a3a&ct=ffffff'/></a>
    </div>
  </footer>
</body>

</html>